Device App Docs

tamigoodall opened the pull request:
Created doc 1 of 3

<!---
Notes:
- Your PR should conform to our [Contributing Guidelines](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md)
- If applicable, add details to the [update feed](https://github.com/auth0/docs/tree/master/updates)
- Make sure your PR gets deployed to a Heroku [Review App](https://github.com/auth0/docs/blob/master/CONTRIBUTING.md#review-apps) without errors
- Please include a short description
- If your PR is a work in progress, title it [DO NOT MERGE]
--->

tamigoodall committed:
Create device-app-auth-flow.md

Created doc 1 of 3

tamigoodall committed:
Added draft flow diagrams and how tos

Added draft flow diagrams and how to doc

tamigoodall committed:
Added more content

Rearranged diagrams
Added notes from source docs
Added links to Logs (TBD)
Updated index.md file under flows

rkhoriander committed:
Create device microsite, rework flow doc, new image for microsite

rkhoriander committed:
merge master

rkhoriander committed:
Device Auth Initial Docs for review

rkhoriander committed:
Fix typo

rkhoriander committed:
Merge branch 'master' into device-apps-docs

rkhoriander committed:
Change microsite name per feedback

rkhoriander committed:
Merge branch 'master' into device-apps-docs

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+### Device Flow
+
+1. The user starts the app on the device.
+2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/device/code** endpoint).
```
```suggestion
2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/oauth/device/code** endpoint).
```

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+
+1. The user starts the app on the device.
+2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/device/code** endpoint).
+3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
```
```suggestion
3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `verification_uri_complete`, `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
```

ziluvatar commented on articles/microsites/call-api/call-api-device.md:
```diff
@@ -0,0 +1,79 @@
+
+1. If the app is not already authorized, our SDK redirects the user to your Auth0 Authorization Server.
+2. Auth0 responds by asking your user to use their laptop or smartphone to visit a specific URL and provide an activation code.
+3. Your device app begins to poll your Auth0 Authorization Server for an Access Token.
```
Steps 1 and 2 are not correct. Check steps 1-4 on `flows/concepts/device-auth.md` "How it works".

ziluvatar commented on articles/microsites/call-api/call-api-device.md:
```diff
@@ -0,0 +1,79 @@
+5. Auth0 responds to your device app with an Access Token.
+6. The Access Token can be used to call your API and retrieve requested data.
+
+For devices, Auth0 uses the [Device Authorization Flow](/flows/concepts/device-auth).
```
Not clear to me, is this saying that the steps above are not for devices?

rkhoriander replied:
No, it's just giving the flow the name the spec uses and linking to the more in-depth description of the flow itself. The microsite is the highest-level look at the job-to-be-done. The flow contains the details about the flow. All of our microsites are written the same way.

ziluvatar commented on articles/microsites/call-api/call-api-device.md:
```diff
@@ -0,0 +1,79 @@
+
+For devices, Auth0 uses the [Device Authorization Flow](/flows/concepts/device-auth).
+
+<img src="/media/articles/microsites/overview-flow-call-api-device-app.png" alt="Flow Overview for Device Apps" width="100%">
```
Not sure if it was on purpose to get the urls shorter, the endpoints are `/oauth/device/code` and `/oauth/token` (`/activate` is fine)

ziluvatar replied:
(just unresolving so we know this is left)

rkhoriander replied:
Just updated these and pushed this, so resolving. :)

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+---
+# Device Authorization Flow
+
+With devices that connect wirelessly to the internet, rather than authenticate the user directly, the device asks the user to go to a link on their computer or smartphone and authorize the device. This avoids a poor user experience for devices that do not have an easy way to enter text. To do this, device apps use the Device Authorization Flow (drafted in [OAuth 2.0](https://tools.ietf.org/html/d ...
```
🤔 It should not matter how the device is connected to internet, it could use ethernet. Can't we just say "With input-constrained devices that connect to the internet, rather than..."?

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+
+The Device Authorization Flow contains two different paths; one occurs on the device requesting authorization and the other occurs in a browser. The browser flow path, wherein a device code is bound to the session in the browser, occurs in parallel to part of the device flow path.
+
+![Device Authorization Sequence](/media/articles/flows/concepts/auth-sequence-device-auth.png)
```
Use `/oauth/device/code` and `/oauth/token`

ziluvatar replied:
Should we mention that the user needs to confirm in step (1) in browser? maybe it's too detailed 🤔 
In steps 4 for browser, since the device path left the authorization server on "Determine whether the user authorized the device", should it show "Authorize device" message? That way the condition whether the device path got stopped gets satisfied.

rkhoriander replied:
@ziluvatar, thanks for pointing out the problem with the endpoint names. This was the language we used on all the other flow docs (/token) and in some other areas of the site, so I followed suit, but I see that it is inaccurate. I'll fix this here and open a separate PR to take care of the other site areas. Great feedback!

rkhoriander replied:
@ziluvatar, we saw step 3 of the browser flow (the "consent" piece) as being the authorization, which is how it is represented in our other flows. That said, we did struggle with how to show the two flows visually.

rkhoriander replied:
I have a request in to graphics to update the image with the proper endpoint names.

ziluvatar replied:
> we saw step 3 of the browser flow (the "consent" piece) as being the authorization, which is how it is represented in our other flows. 

Ok... consent is optional (and authenticate too in case of using SSO, but that's another story), in any case in step 4, for me, it is the authorization server (Auth0 Tenant) the one that does something, and it does not generate an access token, it activates or marks as authorized the device.
Maybe it is like that in other flows because Auth0 Tenant would return the tokens to the client in next step, however, with two flows here: either the generation goes to the device flow or we should rename the step for the browser flow.

rkhoriander replied:
Makes sense. Ok, I've asked Stacy to change step 4 in the Browser Flow to "Marks device as authorized". Will leave this open until I get the graphic back and have added it to the PR.

rkhoriander replied:
@ziluvatar , diagram has been updated.

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/device/code** endpoint).
+3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
+4.  The device app asks the user to activate using their computer or smartphone and displays the `verification_uri` and `user_code`. (Instead of the user code, the app could also display a QR code or shortened URL with embedded user code.)
+5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.
```
```suggestion
5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/oauth/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.
```

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+
+### Browser Flow
+
+1. The user visits the `verification_URL` on their computer. If the `user_code` is not embedded in the `verification_URL`, then the user enters the `user_code`.
```
```suggestion
1. The user visits the `verification_uri` on their computer. If the `user_code` is not embedded in the `verification_uri`, then the user enters the `user_code`.
```

ziluvatar replied:
Should we mention that the user needs to confirm the device that is being activated is displaying the user code?

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+### Browser Flow
+
+1. The user visits the `verification_URL` on their computer. If the `user_code` is not embedded in the `verification_URL`, then the user enters the `user_code`.
+2. Your Auth0 Authorization Server redirects the user to the login and authorization prompt.
```
```suggestion
2. Auth0 Authorization Server redirects the user to the login and consent prompt if needed.
```

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+
+The easiest way to implement the Device Authorization Flow is to follow our tutorial: [Add Authorization Using the Device Authorization Flow](/flows/guides/device-auth/add-auth-device-auth).
+
+## Revoking Device Authorizations
```
Not sure if this is accurate. We have "Devices" and we have "Authorized Applications", if the user authorized the device app but the device app didn't ask for a Refresh Token then admins will not see anything on "Devices", however, the grant remains on the "Authorized Applications" tab.
Maybe it's more confusing to add that distinction, or we can simply modify the title to "Revoking Device Offline Access", meaning like if the device app wants to access the API later, they will need to run the device flow again.

ziluvatar commented on articles/flows/guides/device-auth/call-api-device-auth.md:
```diff
@@ -0,0 +1,88 @@
+---
+title: Call API Using Device Authorization Flow
```
Should we modify the urls to `device-flow` instead of `device-auth`? Not strong opinion just, it is commonly known as device flow maybe it helps for SEO.

rkhoriander replied:
We usually make the URL close to the name in the spec, so if they decide to come out with another flow related to devices, we're able to distinguish easily between the two. I did change the keyword in the "topics" metadata, though.

ziluvatar commented on articles/flows/guides/device-auth/call-api-device-auth.md:
```diff
@@ -0,0 +1,88 @@
+
+* [Register the Application with Auth0](/dashboard/guides/applications/register-app-native). 
+  * Select an **Application Type** of **Native**.
+  * Add an **Allowed Callback URL** of **`YOUR_CALLBACK_URL`**. Your callback URL format will vary depending on your platform. For details about the format for your platform, see our [Native/Mobile Quickstarts](/quickstart/native).
```
This is not necessary/used for device flow.

ziluvatar commented on articles/flows/guides/device-auth/call-api-device-auth.md:
```diff
@@ -0,0 +1,88 @@
+## Steps
+
+1. [Request device code](#request-device-code): Request a device code that the user can use to authorize the device.
+2. [Request Tokens](#request-tokens): 
```
I guess steps 1 and 2 are not done yet, right?

ziluvatar commented on articles/flows/guides/device-auth/includes/authorize-user-call-api.md:
```diff
@@ -0,0 +1,71 @@
+## Authorize the User
```
I guess this doc is wip, right?

ziluvatar commented on articles/flows/guides/device-auth/includes/refresh-tokens.md:
```diff
@@ -0,0 +1,51 @@
+You have already received a [Refresh Token](/tokens/refresh-token/current) if you've been following this tutorial and completed the following:
+
+* configured your API to allow offline access
+* included the `offline_access` scope when you initiated the authentication request through the [authorize](/api/authentication/reference#authorize-application) endpoint
```
We need to add the endpoint to the explorer https://auth0team.atlassian.net/browse/IPS-378 (maybe iam-protocols-sessions can do that with your help)

rkhoriander replied:
Done.

ziluvatar commented on articles/flows/guides/device-auth/includes/request-device-code.md:
```diff
@@ -0,0 +1,70 @@
+
+Note that when requesting a device code, you:
+
+- must include an audience parameter
```
It depends if the app wants an access token for an api or if the app simply wants an access token to get information about the logged user (oidc).
Api => must
Profile => not needed

ziluvatar replied:
Marking as unresolved, I still can see it

panva replied:
it's below

```
::: note
 If your app wants an Access Token only to retrieve info about the authenticated user, then no audience parameter is required.
:::
```

rkhoriander replied:
Technically, this entire guide is about calling a custom API using the flow (which is why configuring the API is a prerequisite) and focuses on getting an Access Token, as that's how we've written the other tutorials we have for other flows. For the moment, I have changed the intro sentence to "to call a custom API".

We could expand this, of course, but we'd need to add more detail about including openid and the /userinfo endpoint, etc. Do you suggest we switch the aim of this particular job-to-be-done? (And if so, do you think it's something we should do for this release, or iterate on?)

Edit: I went ahead and added some barebones info about including openid and using /userinfo in the "scopes" section of the table. Is that a happy medium, you think?

ziluvatar replied:
Sure 👍 we are good.

ziluvatar commented on articles/tokens/refresh-token/current/index.md:
```diff
@@ -184,6 +184,8 @@ You should only ask for a new token if the Access Token has expired or you want
 
 Since Refresh Tokens never expire, it is essential to be able to revoke them in case they get compromised.
 
+For the Device Authorization Flow, the only wy to revoke a device authorization is to revoke the Refresh Token assigned to the device. To learn how, see [Remove Devices from Users](/dashboard/guides/users/remove-devices-users).
```
```suggestion
For the Device Authorization Flow, the only way to revoke a device authorization is to revoke the Refresh Token assigned to the device. To learn how, see [Remove Devices from Users](/dashboard/guides/users/remove-devices-users).
```
You can revoke device authorization on "Authorized Applications" (it will remove the refresh tokens as well).

rkhoriander committed:
Respond to feedback; finalize tutorial

rkhoriander commented:
Main changes:

- https://docs-content-staging-pr-7461.herokuapp.com/docs/microsites/call-api/call-api-device
- https://docs-content-staging-pr-7461.herokuapp.com/docs/flows/concepts/device-auth
- https://docs-content-staging-pr-7461.herokuapp.com/docs/flows/guides/device-auth/call-api-device-auth

ziluvatar commented on articles/flows/guides/device-auth/call-api-device-auth.md:
```diff
@@ -0,0 +1,82 @@
+## Steps
+
+1. [Request device code](#request-device-code) (Device Flow): Request a device code that the user can use to authorize the device.
+2. [Request device authorization](#request-device-authorization) (Device Flow): Request that the user authorize the device using their laptop or smartphone.
```
It might be confusing with the spec, since "Request device code" is a "Device Authorization Request" in the spec, however, "Request device authorization" is "user interaction" in spec or "Device Activation" in our UI.
(It's fine if you leave it like this, I just wanted to raise it)

rkhoriander replied:
Changed to "Request device activation" to match our UI.

ziluvatar commented on articles/flows/guides/device-auth/includes/request-tokens.md:
```diff
@@ -0,0 +1,74 @@
+
+### Responses
+
+While you wait for the user to authorize the device, you may receive a few different `HTTP 400` responses:
```
```suggestion
While you wait for the user to authorize the device, you may receive a few different `HTTP 4xx` responses:
```
Each error has more or less different 4xx error codes. Dismiss this if you knew and `HTTP 400` is to avoid confusion.

rkhoriander replied:
I knew, but I think it's good to be more precise here. Accepting your suggestion.

ziluvatar commented on articles/flows/guides/device-auth/includes/request-tokens.md:
```diff
@@ -0,0 +1,74 @@
+
+#### Access Denied 
+
+Finally, if the user refuses to authorize the device, you will receive:
```
or a rule or the authorization server denied the transaction

ziluvatar commented on articles/flows/guides/device-auth/includes/user-authorization.md:
```diff
@@ -0,0 +1,24 @@
+
+The user will either scan the QR code, or else will open the activation page and enter the user code:
+
+![Enter User Code](/media/articles/flows/guides/device-auth/enter-user-code.png)
```
Just a suggestion, maybe we should use the tenant from the playground or any other tenant instead of the one from sandrino

rkhoriander replied:
Done.

ziluvatar commented on articles/flows/guides/device-auth/includes/receive-tokens.md:
```diff
@@ -0,0 +1,26 @@
+
+While the user has been authenticating and authorizing the device, the device app has continued to poll the token URL to request an Access Token. 
+
+Once the user has suucessfully authorized the device, you'll receive an `HTTP 200` response with a payload containing `access_token`, `refresh_token`, and `token_type` values:
```
optionally it could get an `id_token` too.

(edit by @panva: optionally if the device authorization request included `openid` scope it will get an `id_token` as well)

rkhoriander replied:
I changed the "request device code" section ("scope" parameter in table) and added that you need to add openid to get the ID Token. Between that and the added "(optionally)" for ID Token and Access Token, I think this is done.

ziluvatar commented on articles/flows/guides/device-auth/includes/sample-use-cases-call-api.md:
```diff
@@ -0,0 +1,43 @@
+## Sample Use Cases
+
+### Customize Tokens
```
Does this use case belong to here?

rkhoriander replied:
Removed.

ziluvatar commented on articles/flows/guides/device-auth/includes/troubleshooting.md:
```diff
@@ -0,0 +1,24 @@
+
+| Code       | Name | Description |
+|------------|------|-------------|
+| `limit_dc` | Too Many Invalid Device Codes	| |
```
Not used, it can be removed

ziluvatar commented on articles/flows/guides/device-auth/includes/troubleshooting.md:
```diff
@@ -0,0 +1,24 @@
+| `limit_dc` | Too Many Invalid Device Codes	| |
+| `fdeaz`    | Failed device authorization request | |	
+| `fdeac`    | Failed device activation | |	
+| `fede`     | Failed Exchange | Device Code for Access Token |
```
There is another one: 

fdecc | User canceled the device confirmation.
-- | --

ziluvatar commented on articles/flows/guides/device-auth/includes/troubleshooting.md:
```diff
@@ -0,0 +1,24 @@
+* Non-OIDC-conformant Applications
+* Federated Applications
+* Application types where the authentication method is not set to **None**
+* Applications created through Dynamic Client Registration.
```
More:
* Only available for "Native" applications.
* Not fully compatible with connections using Auth0 developer keys.

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+
+### Browser Flow
+
+1. The user visits the `verification_uri` on their computer and confirms that the device that is being activated is displaying the `user_code`. If the `user_code` is not embedded in the `verification_uri`, then the user enters the `user_code`.
```
```suggestion
1. The user visits the `verification_uri` on their computer, enters the `user_code` and confirms that the device that is being activated is displaying the `user_code`. If the user visits the `verification_uri_complete` by any other mechanism (such as by scanning a QR code), only the device confirmation will be needed.
```

ziluvatar commented:
Can we add the new grant type to https://docs-content-staging-pr-7461.herokuapp.com/docs/applications/reference/grant-types-available too?

rkhoriander committed:
Merge master

rkhoriander committed:
Update device flow and authentication sequence images

ziluvatar committed:
Update articles/flows/concepts/device-auth.md

Co-Authored-By: rkhoriander <rachel.khoriander@auth0.com>

ziluvatar committed:
Update articles/flows/guides/device-auth/includes/request-tokens.md

Co-Authored-By: rkhoriander <rachel.khoriander@auth0.com>

panva commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/oauth/device/code** endpoint).
+3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `verification_uri_complete` `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
+4.  The device app asks the user to activate using their computer or smartphone and displays the `verification_uri` and `user_code`. (Instead of the user code, the app could also display a QR code or shortened URL with embedded user code.)
+5. The device app begins polling your Auth0 Authorization Server for an Access Token (**/oauth/token** endpoint) using the time period specified by `interval`. The device app continues polling until either the user completes the browser flow path or the user code expires.
```
based on our own mishaps in client implementation we should note that each interval should start counting after receipt of the last request's response to avoid `slow_down` errors due to network latency

panva commented on articles/flows/guides/device-auth/includes/request-device-code.md:
```diff
@@ -0,0 +1,73 @@
+* 20 characters
+
+**Maximum expiration time**:
+* 10 minutes
```
We do not allow expiration time to be adjusted.

ziluvatar replied:
Also, it is 15 minutes: https://github.com/auth0/auth0-server/blob/master/lib/config/device_flow.js#L22 it seems the description from the task was wrong :(

panva commented on articles/flows/guides/device-auth/includes/request-device-code.md:
```diff
@@ -0,0 +1,73 @@
+* Numbers: 9 characters
+
+**Maximum length**: 
+* 20 characters
```
20 characters including spaces and hyphens (separators added for readability)

panva commented on articles/flows/guides/device-auth/includes/request-tokens.md:
```diff
@@ -0,0 +1,74 @@
+
+#### Expired Token
+
+The user has not authorized the device quickly enough, so the `device_code` has expired. You should wait for the user to restart the device authorization flow to avoid unnecessary polling.
```
Change

> You should wait for the user to restart the device authorization flow to avoid unnecessary polling.

to

> Your application should notify the user of the flow being expired and prompt to re-initiate the flow.

+ note, `expired_token` will be returned exactly once, afterwards `invalid_grant` will be returned, the device MUST stop polling.

rkhoriander committed:
Merge branch 'master' into device-apps-docs

rkhoriander committed:
Text changes in response to review feedback

rkhoriander committed:
Updated text in response to most recent feedback

rkhoriander committed:
Replaced lost note

rkhoriander committed:
Added info about openid and /userinfo endpoint

rkhoriander committed:
Added grant to grant ref docs

rkhoriander committed:
Updated browser flow images from playground

rkhoriander committed:
Final troubleshooting/proofreading corrections

ziluvatar commented on articles/flows/guides/device-auth/includes/troubleshooting.md:
```diff
@@ -0,0 +1,27 @@
+* Federated Applications
+* Applications created through [Dynamic Client Registration](/api-auth/dynamic-client-registration)
+* [Social Connections](/connections) using [Auth0 developer keys](/connections/social/devkeys)
+* Custom query string parameters
```
Maybe we can clarify:
```suggestion
* Query string parameters to be accessed from hosted login page or rules.
```

rkhoriander committed:
Update articles/flows/guides/device-auth/includes/troubleshooting.md

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
Update troubleshooting.md

rkhoriander committed:
Corrections to limitations

rkhoriander committed:
Updated last step of browser flow in auth sequence diagram

ziluvatar commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+  "link": "#device-auth"
+}) %>
+
+This is the OAuth 2.0 grant that input-constrained devices use to access an API. Poll this endpoint using the interval returned with your [device code](/api/authentication?http#device-code) to directly request an Access Token using the application's credentials (a Client ID) and a device code.
```
Is this link right? it didn't show me anything

rkhoriander replied:
So... intriguingly, this is a piece of functionality that seems to try to encase the GET label in an <a> tag with the `href=` being this "link". But the functionality obviously doesn't work on any of the endpoints. It isn't breaking anything either, though, so for now, we will leave this as is until we dig into this functionality further.

ziluvatar replied:
Just to see if I understood, you want to merge it and fix it later or you want to leave the conversation unresolved until you dig into the issue?

rkhoriander replied:
I misunderstood this issue. I was focusing on the badge code in the "include", but I see that you mean the link below that (just above your comment). Will be fixed in the next push.

ziluvatar commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+|:-----------------|:------------|
+| `grant_type` <br/><span class="label label-danger">Required</span> | Denotes the flow you are using. For Device Authorization, use `urn:ietf:params:oauth:grant-type:device_code`. |
+| `client_id` <br/><span class="label label-danger">Required</span> | Your application's Client ID. |
+| `device_code` <br/><span class="label label-danger">Required</span> | The device code previously returned from the [/oauth/device/code endpoint](/api/authentication?http#device-code). |
```
🤔 I think this should be `get-device-code`?

ziluvatar commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+Content-Type: application/json
+ { 
+  "error": "authorization_pending",
+  "error_description": "Retryable: We are waiting for the user to take some action; try polling again using the suggested interval."
```
```suggestion
  "error_description": "User has yet to authorize device code."
```

ziluvatar replied:
Maybe you wanted to be more clear and ignore the real message, dismiss then.

ziluvatar commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+Content-Type: application/json
+ { 
+  "error": "slow_down",
+  "error_description": "Retryable: You are polling too fast; slow down and use the suggested interval."
```
```suggestion
  "error_description": "You are polling faster than the specified interval of 5 seconds."
```

ziluvatar replied:
Maybe you wanted to be more clear and ignore the real message, dismiss then.

ziluvatar commented on articles/applications/reference/grant-types-available.md:
```diff
@@ -26,6 +26,7 @@ Various grant types are valid when registering Auth0 Applications. These can be
 | `client_credentials` | [Client Credentials Grant](/flows/concepts/client-credentials) |
 | `password` | [Resource Owner Password Grant](/api-auth/grant/password) |
 | `refresh_token` | [Use a Refresh Token](/tokens/refresh-token/current#use-a-refresh-token) |
+|  `device_code` | [Device Authorization Grant](/flows/concepts/device-auth) |
```
```suggestion
|  `urn:ietf:params:oauth:grant-type:device_code` | [Device Authorization Grant](/flows/concepts/device-auth) |
```

ziluvatar commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+1. The user visits the `verification_uri` on their computer, enters the `user_code` and confirms that the device that is being activated is displaying the `user_code`. If the user visits the `verification_uri_complete` by any other mechanism (such as by scanning a QR code), only the device confirmation will be needed.
+2. Your Auth0 Authorization Server redirects the user to the login and consent prompt, if needed.
+3. The user authenticates using one of the configured login options and may see a consent page asking to authorize the device app.
+4. If the `user_code` is valid, when the user consents, your device app is authorized to access the API.
```
```suggestion
4. Your device app is authorized to access the API.
```
(suggestion based on the updated image, because the user code was validated on step 1)

rkhoriander committed:
Update articles/applications/reference/grant-types-available.md

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
Update articles/flows/concepts/device-auth.md

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
Fix slow_down error description

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
Fix authorization_pending error description.

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
merge master

Commit:
[IPS-391] Add a basic quickstarts (link to docs) under "Native" application type

rkhoriander committed:
Fix api link

rkhoriander committed:
Fix bookmark

rkhoriander committed:
Merge branch 'device-apps-docs' of github.com:auth0/docs into device-apps-docs

sandrinodm commented on articles/api/authentication/api-authz/_device-code.md:
````diff
@@ -0,0 +1,79 @@
+
+```JSON
+HTTP/1.1 200 OK
+Content-Type: application/json
````
In this response sample please replace `example.com` with `${account.namespace}`

rkhoriander replied:
Done.

sandrinodm commented on articles/api/authentication/api-authz/_device-code.md:
```diff
@@ -0,0 +1,79 @@
+  "verification_uri":"https://example.com/device",
+  "verification_uri_complete":"https://example.com/device?user_code=WDJB-MJHT",
+  "expires_in":900,
+  "interval":5
```
Can we clarify that interval is in seconds somehow?

rkhoriander replied:
Added this in as a comment. Although technically this shouldn't include comments, since this is a response, it shouldn't cause trouble for users.

sandrinodm commented on articles/api/authentication/api-authz/_get-token.md:
````diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+}
+```
+
+```JSON
````
Can we explain which ones are retryable and which ones note? The authorization_pending and slow_down can be retried, the others not. The other errors mean the transaction failed.

rkhoriander replied:
Added this in as a comment. Although technically this shouldn't include comments, since this is a response, it shouldn't cause trouble for users.

sandrinodm commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+Content-Type: application/json
+ { 
+    "error": <anything else>,
+    "description": "Failure: User cancelled the confirmation prompt or consent page; the code expired; there was an error."
```
`error_description`

sandrinodm commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+HTTP/1.1 400 BAD REQUEST
+Content-Type: application/json
+ { 
+    "error": <anything else>,
```
maybe replace anything else with `access_denied|invalid_grant|...`

sandrinodm commented on articles/api/authentication/api-authz/_get-token.md:
```diff
@@ -390,6 +390,109 @@ This is the OAuth 2.0 grant that highly-trusted apps use to access an API. In th
+| `device_code` <br/><span class="label label-danger">Required</span> | The device code previously returned from the [/oauth/device/code endpoint](/api/authentication?http#get-device-code). |
+
+### Remarks
+- Because you will be polling this endpoint while waiting for the user to go to the verification URL and enter their user code, you will likely receive at least one failure before receiving a successful response. See sample responses for possible responses.
```
Can we clarify that the polling frequency should used the `interval` from the intial response?

sandrinodm commented on articles/dashboard/guides/users/remove-devices-users.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: Remove Devices from Users
```
I don't think the wording "Removing Devices " is correct. We need better wording here. Maybe ask @vibronet for input about this.

IMO you are "signing out from the device" because:

 - You still have your device 🤔 It's not gone or removed.
 - The grant is still there, so the device is still authorized

rkhoriander replied:
Changed to "Unlink Devices" to match Dashboard terminology. For the how-tos, the Dashboard terminology generally drives the titling. In the sections that link to the guide, we call it "forcing device reauthorization".

sandrinodm commented on articles/flows/guides/auth-code-pkce/includes/authorize-user-call-api.md:
```diff
@@ -26,7 +26,7 @@ https://${account.namespace}/authorize?
 #### Parameters
 
-Note that for authorizing a user when calling an API, you:
+Note that for authorizing a user when calling a custom API, you:
```
Why are we adding `custom` here? What is a "non-custom" API then? 🤔

rkhoriander replied:
The docs use the word "custom" to refer to an API that the user is building, as opposed to the Management API or similar. I just updated the wording in the auth-code and auth-code-pkce flows to match what we're doing in the device code flow since it seemed a similar objection could be had there as well.

sandrinodm commented on articles/flows/guides/auth-code/includes/authorize-user-call-api.md:
```diff
@@ -24,7 +24,7 @@ https://${account.namespace}/authorize?
 #### Parameters
 
-Note that for authorizing a user when calling an API, you:
+Note that for authorizing a user when calling a custom API, you:
```
Same, why custom?

rkhoriander replied:
Same as above:
"The docs use the word "custom" to refer to an API that the user is building, as opposed to the Management API or similar. I just updated the wording in the auth-code and auth-code-pkce flows to match what we're doing in the device code flow since it seemed a similar objection could be had there as well."

rkhoriander committed:
Updated to correct doc author

Took this project over after initial page creation, so if there are questions, I'm likely the most able to answer at the moment.

rkhoriander committed:
Update _device-code.md

rkhoriander committed:
Merge branch 'master' into device-apps-docs

rkhoriander committed:
Merge branch 'device-apps-docs' of github.com:auth0/docs into device-apps-docs

rkhoriander committed:
Changes from review feedback

asmclean commented:
I didn't realize this initially, but the `autho-docs` test suite doesn't allow quickstarts to contain duplicate headings, so re-using the guide in its current state causes two tests to fail.  Here's an example of some changes that can be used to get the tests to pass using the guide as a quick start:
https://github.com/asmclean/docs/commit/176a6ac9a04904391cdf2e330dd0cd3a39354bad
@rkhoriander Thoughts?  Do those look like reasonable edits?

rkhoriander commented:
@asmclean, Hi there. Yes, those edits look fine to me. :)

rkhoriander committed:
Merge branch 'device-apps-docs' of github.com:auth0/docs into device-apps-docs

rkhoriander committed:
Added final round of changes from feedback

rkhoriander committed:
Merge branch 'master' into device-apps-docs

Commit:
Fix quickstart validation errors due to duplicate headings

ziluvatar commented on articles/dashboard/guides/users/unlink-user-devices.md:
```diff
@@ -0,0 +1,30 @@
+---
+title: Remove Devices from Users
```
```suggestion
title: Unlink Devices
```
You said in one comment you wanted to rename it to ^ . Adding just in case you modified only the path and not the title.

ziluvatar commented on articles/dashboard/guides/users/unlink-user-devices.md:
```diff
@@ -0,0 +1,30 @@
+
+2. Click the **Devices** view, then click **Unlink** next to the device you want to unlink, and confirm.
+
+
```
Shouldn't we show here an image of the device tab?

ziluvatar replied:
nvm, we can add it once this https://auth0team.atlassian.net/browse/IPS-343 is done if we want.

ziluvatar reviewed (changes requested):
Can we add to the refresh token list on https://docs-content-staging-pr-7461.herokuapp.com/docs/api/authentication#get-token the device authorization flow as well?

rkhoriander committed:
Missing quotes

rkhoriander committed:
Update articles/dashboard/guides/users/unlink-user-devices.md

Co-Authored-By: Eduardo Diaz <eduardo.diaz@auth0.com>

rkhoriander committed:
Update metadata

rkhoriander committed:
Add device flow to refresh token list

sandrinodm commented on articles/api/authentication/api-authz/_device-code.md:
```diff
@@ -0,0 +1,79 @@
+  "device_code":"GmRh...k9eS",
+  "user_code":"WDJB-MJHT",
+  "verification_uri":"https://${account.namespace}/device",
+  "verification_uri_complete":"${account.namespace}/device?user_code=WDJB-MJHT",
```
HTTPS is missing before this url

sandrinodm commented on articles/api/authentication/api-authz/_device-code.md:
```diff
@@ -0,0 +1,79 @@
+Content-Type: application/json
+{
+  "device_code":"GmRh...k9eS",
+  "user_code":"WDJB-MJHT",
```
Are we going to have a table explaining what these response values mean?

Eg:

 - device_code: Your application will use this code when polling the /oauth/token endpoint
 - user_code: The code your end user will need to fill in. You typically render this in your page asking them to go to the verification_uri and fill in the code
 - verification_uri: The URI where a user can enter the code
 - verification_uri_complete: The URI which also includes the user_code. Useful if you want to open a browser with the code already filled in or you need to generate a QR code.
 - expires_in: Expiration of the codes (in seconds)
 - interval: Polling interval in seconds for the client to call the /oauth/token endpoint

rkhoriander replied:
Strictly following the format of our API ref, this info isn't typically included. I can see where it is useful, though, both generally and particularly for this flow, so I am adding it for this section. I'll add a ticket as part of our API ref rework to consider adding response values for other endpoints.

sandrinodm commented on articles/flows/concepts/device-auth.md:
```diff
@@ -0,0 +1,60 @@
+1. The user starts the app on the device.
+2. The device app requests authorization from the Auth0 Authorization Server using its Client ID (**/oauth/device/code** endpoint).
+3. The Auth0 Authorization Server responds with a `device_code`, `user_code`, `verification_uri`, `verification_uri_complete` `expires_in` (lifetime in seconds for `device_code` and `user_code`), and polling `interval`.
+4.  The device app asks the user to activate using their computer or smartphone and displays the `verification_uri` and `user_code`. (Instead of the user code, the app could also display a QR code or shortened URL with embedded user code.)
```
Explain how `verification_uri_complete` can be used for generating a QR code. Additionally, also explain that for a native app running on your computer, the url can be used to open a browser.

sandrinodm commented on articles/flows/guides/device-auth/includes/request-device-code.md:
```diff
@@ -0,0 +1,73 @@
+## Request Device Code
+
```
This looks like a repeat of `articles/api/authentication/api-authz/_device-code.md`. The other article has much more code samples it seems ,but is lacking an explanation of the response.

Why do we have this duplicate content? Is there a chance we can unify this?

rkhoriander replied:
This is an include file that serves as part of a tutorial. The other doc you mention is the API reference, which mimics the Swagger-generated Mgmt API ref. 

For every flow, we include the API reference as well as a microsite (for entry-level users), a flow (with more mid-level detail), and a tutorial (with suggestions to use our quickstarts/SDKs, where available, and a basic "roll-your-own" implementation). For the moment, we don't intend to unify these items as they serve different purposes for their intended audiences.

rkhoriander replied:
As we improve our API refs, though, I hope we can bring these more closely in line and/or unify this info.

rkhoriander replied:
For reference, here is the tutorial:
https://docs-content-staging-pr-7461.herokuapp.com/docs/flows/guides/device-auth/call-api-device-auth

Here is the API ref:
https://docs-content-staging-pr-7461.herokuapp.com/docs/api/authentication#device-authorization-flow

sandrinodm commented on articles/flows/guides/device-auth/includes/request-tokens.md:
```diff
@@ -0,0 +1,85 @@
+| `device_code`   | The `device_code` retrieved in the previous step of this tutorial. |
+| `client_id`     | Your application's Client ID. You can find this value in your [Application Settings](${manage_url}/#/Applications/${account.clientId}/settings). |
+
+### Token Responses
```
This seems like a repeat of `articles/api/authentication/api-authz/_get-token.md`. Why do we have duplicate content here? This seems to be better formatted and explains things better.

rkhoriander replied:
Same as my prior answer--this is part of a tutorial whereas the other is the API reference.

rkhoriander replied:
But thank you. This doc is my writing/style, and the other isn't. :)

rkhoriander committed:
Respond to Sandrino feedback add log entries

rkhoriander committed:
Add dashboard screenshots

rkhoriander committed:
merge master

rkhoriander commented:
@sandrinodimattia , can you please review the convos/edits and approve this PR so we can release this tomorrow?
